stanford_soe_r25.pages.inc

```diff
@@ -0,0 +1,994 @@
+  ) {
+    // get the list of email addresses for the security group that can do the approvals
+    $approver_list =
+      _stanford_soe_r25_security_group_emails($form_state['storage']['stanford_soe_r25_booking']['room']['approver_secgroup_id']);
+    if (!empty($approver_list)) {
+      // for each approver in the security group, add their email address to the list and
+      // create a "to do task" XML snippet with their 25Live id and the event information
+      // to be added to the request XML.
+      $todo_str = file_get_contents(drupal_get_path('module', 'stanford_soe_r25') . '/stanford_soe_r25_reserve_todo.xml');
+      foreach ($approver_list as $key => $value) {
+        $todo_temp = str_replace('[r25_start_date_time]', $form_state['storage']['stanford_soe_r25_booking']['dates']['start'], $todo_str);
+        $todo_temp = str_replace('[r25_approver_id]', $key, $todo_temp);
+        $todo_temp = str_replace('[r25_credential_id]', variable_get('stanford_soe_r25_credential_contact_id', ''), $todo_temp);
+        $todo_insert .= $todo_temp;
+        if (!empty($mail_list)) {
+          $mail_list .= ', ';
+        }
+        $mail_list .= $value;
+      }
+    }
+  }
+
+  // if there are 25Live custom attributes associated with this event, add the XML snippet for each attribute
+  // with its id, type, and value to the request XML.
+  $attr_insert = '';
+  $attr_str = file_get_contents(drupal_get_path('module', 'stanford_soe_r25') . '/stanford_soe_r25_reserve_attr.xml');
+  $room = $form_state['storage']['stanford_soe_r25_booking']['room'];
+
+  if (!empty($room['event_attribute_fields'])) {
+    foreach ($room['event_attribute_fields'] as $key => $value) {
+      if (isset($form_state['values']['stanford_soe_r25_booking_attr' . $key])) {
+        $attr_temp = str_replace('[r25_attr_id]', $key, $attr_str);
+        $attr_temp = str_replace('[r25_attr_type]', $value['type'], $attr_temp);
+        $attr_temp = str_replace('[r25_attr_value]', $form_state['values']['stanford_soe_r25_booking_attr' . $key], $attr_temp);
+        $attr_insert .= $attr_temp;
+      }
+    }
+  }
+
+  // if there is a 25Live custom attribute we've defined for contact information, add the XML snippet for it to the request
+  if (!empty($room['contact_attr_field'])) {
+    foreach ($room['contact_attr_field'] as $key => $value) {
+      if (!empty($form_state['values']['stanford_soe_r25_contact_' . $key])) {
+        $attr_temp = str_replace('[r25_attr_id]', $key, $attr_str);
+        $attr_temp = str_replace('[r25_attr_type]', $value['type'], $attr_temp);
+        $attr_temp = str_replace('[r25_attr_value]', $form_state['values']['stanford_soe_r25_contact_' . $key], $attr_temp);
+        $attr_insert .= $attr_temp;
+      }
+    }
+  }
+
+  // get the XML template for creating an event and replace tokens with data for this reservation
+  $event_state = $event_state - 1;
+  $xml_file = '/stanford_soe_r25_reserve.xml';
```

Comment:
instead of all the `str_replace` for the tokens, you could simply have a xml array, then convert the array to xml.